lib/Sema/CSSimplify.cpp

```diff
@@ -6080,10 +6090,33 @@ ConstraintSystem::simplifyApplicableFnConstraint(
   ConstraintLocatorBuilder outerLocator =
     getConstraintLocator(anchor, parts, locator.getSummaryFlags());
 
-  // Before stripping optional types, save original type for handling
-  // @dynamicCallable applications. This supports the fringe case where
-  // `Optional` itself is extended with @dynamicCallable functionality.
-  auto origType2 = desugar2;
+  // Handle applications of types with `callAsFunction` methods.
+  // Do this before stripping optional types below, when `shouldAttemptFixes()`
+  // is true.
+  auto hasCallAsFunctionMethods =
+      desugar2->mayHaveMembers() &&
+      llvm::any_of(lookupMember(desugar2, DeclName(ctx.Id_callAsFunction)),
+                   [](LookupResultEntry entry) {
+                     return isa<FuncDecl>(entry.getValueDecl());
+                   });
+  if (hasCallAsFunctionMethods) {
+    auto memberLoc = getConstraintLocator(
+        outerLocator.withPathElement(ConstraintLocator::Member));
+    // Add a `callAsFunction` member constraint, binding the member type to a
+    // type variable.
+    auto memberTy = createTypeVariable(memberLoc, TVO_CanBindToLValue |
+                                                  TVO_CanBindToNoEscape |
+                                                  TVO_CanBindToInOut);
+    addValueMemberConstraint(origType2, DeclName(ctx.Id_callAsFunction),
+                             memberTy, DC, FunctionRefKind::SingleApply,
+                             /*outerAlternatives*/ {}, locator);
```

Comment:
Might make sense to add a `TODO` in case we'd ever want to get a static variant of `callAsFunction`...

Reply:
I see - would you like the `TODO` comment to say anything more than ``TODO: Revisit this if `static func callAsFunction` is to be supported.``?

I'm also curious about what constraint changes are necessary to support `static func callAsFunction`.

Reply:
This is about `SingleApply` vs. `DoubleApply` I believe `originalTy2` would already be a metatype value so nothing would have to be done there.

Reply:
I see, that makes sense. (I guess static functions need `DoubleApply` due to metatype currying.)

Reply:
Todo hint added in https://github.com/apple/swift/pull/24299/commits/cebbd6335dab7e0bbf3e5c0dff9c445588e01f81.